hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java

```diff
@@ -1339,7 +1340,14 @@ private static List<Tuple3<String, String, Boolean>> fetchPartitionFileInfoTripl
    * @param recordKey record key for which the file group index is looked up for.
    * @return An integer hash of the given string
    */
-  public static int mapRecordKeyToFileGroupIndex(String recordKey, int numFileGroups) {
+  public static int mapRecordKeyToFileGroupIndex(String recordKey, int numFileGroups, String partitionName) {
+    if (MetadataPartitionType.SECONDARY_INDEX.isPartitionType(partitionName) && recordKey.contains(SECONDARY_INDEX_RECORD_KEY_SEPARATOR)) {
+      return mapRecordKeyToFileGroupIndex(SecondaryIndexKeyUtils.getSecondaryKeyFromSecondaryIndexKey(recordKey), numFileGroups);
+    }
+    return mapRecordKeyToFileGroupIndex(recordKey, numFileGroups);
```

Comment:
We should guard this with write table version, i.e., the new logic should only take effect for the new table version 9.